compiler/rustc_mir_transform/src/jump_threading.rs

```diff
@@ -509,6 +509,13 @@ impl<'tcx, 'a> TOFinder<'tcx, 'a> {
                     BinOp::Ne => ScalarInt::FALSE,
                     _ => return None,
                 };
+                if value.const_.ty().is_floating_point() {
+                    // Floating point equality does not follow bit-patterns.
+                    // -0.0 and NaN both have special rules for equality,
+                    // and therefore we cannot use integer comparisons for them.
+                    // Avoid handling them, though this could be extended in the future.
+                    return None;
+                }
                 let value = value.const_.normalize(self.tcx, self.param_env).try_to_scalar_int()?;
```

Comment:
While this PR is an easily backportable targeted fix, it might be worth thinking about how to prevent this more generally. It seems highly confusing that a method called `try_to_scalar_int` will return floats, and it doesn't surprise me that this pass was confused by this. I would also expect `try_to_scalar_int` to be OK for comparing bitwise.
Maybe it should be changed to not return anything for floats?

Reply:
Possibly bad suggestion, but maybe we shouldn't use valtrees to represent floats 🤔

Reply:
`try_to_bitwise_repr`?